### ISSUE TITLE
GZIPContentDecoder ignores setUseInputDirectByteBuffers setting and always uses non-direct buffers (causing GC locking)

### ISSUE THREAD

**Issue by jaroslawr** [eclipse/jetty.project#7683]
**Jetty version(s)**

10.0.8

**Java version/vendor** `(use: java -version)`

```
openjdk version "17.0.2" 2022-01-18
OpenJDK Runtime Environment (build 17.0.2+8-86)
OpenJDK 64-Bit Server VM (build 17.0.2+8-86, mixed mode, sharing)
```

**OS type/version**

```
Ubuntu 20.04.2 LTS
```

**Description**

GZIPContentDecoder ignores setUseInputDirectByteBuffers setting and always uses non-direct buffers:

https://github.com/eclipse/jetty.project/blob/jetty-10.0.8/jetty-http/src/main/java/org/eclipse/jetty/http/GZIPContentDecoder.java#L435

This came to my attention because Java Inflater class when used with non-direct buffers will use JNI which in turn can cause GC issues. This is well explained in the README of this project (pure java inflate implementation, which has nothing to do with the issue except for having a good explanation of the problem) https://github.com/google/pjflate:

> This is not an attempt to make it fast. In fact, it's much slower than the Java standard Inflater. This is more like to avoid the GCLocker impact.
>
> JNI code in general is not GC aware, and in order to avoid arrays to be moved by JVM GC, it needs to pin primitive arrays. While pinning, JVM GC is paused. After this pinning is done, JVM can trigger a GC (GCLocker initiated GC). Because no GC is done while this pinning, depending on the heuristics JVM can trigger a full GC. For certain applications, it's more favorable to keep concurrent GCs running to prevent a full GC.
>
> [...]
>
> Starting from Java SE 11, the standard Inflater can take a ByteBuffer. In Open JDK implementation, it checks if it's a direct buffer and if it is, it doesn't lock GC. If GCLocker is a concern and Open JDK 11 or later is used, it's better to use a direct buffer with the standard Inflater. It gives you a better speed than pure Java implementations and you can avoid the GCLocker issue.

**How to reproduce?**

https://github.com/eclipse/jetty.project/blob/jetty-10.0.8/jetty-http/src/main/java/org/eclipse/jetty/http/GZIPContentDecoder.java#L435

**Comment by sbordet** [comment]
@lachlan-roberts would it be possible to use direct buffers in Jetty 10+ since we can leverage the new APIs?

**Comment by jaroslawr** [comment]
@lachlan-roberts I created a fix for this, partially based on your old commit https://github.com/eclipse/jetty.project/commit/7a04b3eb19820c66b65d026cfd972a49689dc582 that you later partially reverted in https://github.com/eclipse/jetty.project/commit/2f23c5936e0159d345a93035d7ef14be87516efd. I added back the reverted code from `7a04b3e` in GzipHttpOutputInterceptor, but conditionally on direct buffers being enabled in HttpConfiguration. When used with direct buffers, Java CRC32 class does not make a copy of the input byte buffer, which was the reason for the revert, if I understand right.
